circuits/src/zk_circuits/valid_relayer_fee_settlement.rs

```diff
@@ -415,3 +415,589 @@ where
+    }
+
+    // -------------------------
+    // | Invalid Authorization |
+    // -------------------------
+
+    /// Test the case in which the prover provides an invalid decryption key
+    #[test]
+    fn test_invalid_decryption_key() {
+        let mut rng = thread_rng();
+        let (sender_wallet, recipient_wallet) = get_initial_wallets();
+        let (statement, mut witness) = create_witness_statement(&sender_wallet, &recipient_wallet);
+
+        witness.recipient_decryption_key = DecryptionKey::random(&mut rng);
+        assert!(!check_constraints_satisfied(&statement, &witness));
+    }
+
+    /// Test the case in which the recipient's root key does not match the one
+    /// in the statement
+    #[test]
+    fn test_invalid_root_key() {
+        let (sender_wallet, recipient_wallet) = get_initial_wallets();
+        let (mut statement, witness) = create_witness_statement(&sender_wallet, &recipient_wallet);
+
+        // Zero out the signing key
+        let mut zero_iter = std::iter::repeat(Scalar::zero());
+        statement.recipient_pk_root = PublicSigningKey::from_scalars(&mut zero_iter);
+        assert!(!check_constraints_satisfied(&statement, &witness));
+    }
+
+    // -------------------
+    // | Invalid Reblind |
+    // -------------------
+
+    /// Tests the case in which a share of the sender's wallet is
+    /// modified
+    #[test]
+    fn test_invalid_sender_reblind__random_modification() {
+        let mut rng = thread_rng();
+        let (sender_wallet, recipient_wallet) = get_initial_wallets();
+        let (mut statement, mut witness) =
+            create_witness_statement(&sender_wallet, &recipient_wallet);
+
+        // Choose a point modification
+        let shares_len = WalletShare::<MAX_BALANCES, MAX_ORDERS>::NUM_SCALARS;
+        let modification_idx = rng.gen_range(0..shares_len);
+        let modification = Scalar::random(&mut rng);
+
+        // Modify the private shares
+        let mut private_shares = witness.sender_updated_private_shares.to_scalars();
+        private_shares[modification_idx] += modification;
+        witness.sender_updated_private_shares =
+            WalletShare::from_scalars(&mut private_shares.into_iter());
+        statement.sender_wallet_commitment =
+            compute_wallet_private_share_commitment(&witness.sender_updated_private_shares);
+
+        // Modify the public shares in the opposite way, so that the wallet remains
+        // intact
+        let mut public_shares = statement.sender_updated_public_shares.to_scalars();
+        public_shares[modification_idx] -= modification;
+        statement.sender_updated_public_shares =
+            WalletShare::from_scalars(&mut public_shares.into_iter());
+
+        assert!(!check_constraints_satisfied(&statement, &witness));
+    }
+
+    /// Tests the case in which the public blinder share is incorrect
+    #[test]
+    fn test_invalid_public_blinder_share() {
```

Comment:
do we also want a test case for incorrect public secret shares, to test them being implicitly constrained by the state transition? that checks out to me mentally, but a test may be good to confirm

Reply:
that is what the test above this does